### PR TITLE
implementation of entropy algorithm.

### PR DESCRIPTION
**Pull request by beqakd**
Wrote implementation of entropy algorithm.

* [x] Add an algorithm? #2109 

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by beqakd** [comment]
This fail is not from my PR!

File: maths/entropy.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    This method takes path and two dict as argument
+    and than calculates entropy of them.
+    :param my_dic_fir:
+    :param my_dic_sec:
+    :param path_dest:
+    :return:
+    """
+    f = open(path_dest, mode="w+", encoding="utf-8")
+
+    my_alphas = create_alpha_array()  # create our alpha
```
Comment (cclauss) [inline maths/entropy.py:33]:
```suggestion
    my_alphas = list(' ' + ascii_lowercase)
```

File: maths/entropy.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
+"""
+    Implementation of entropy of information
+    https://en.wikipedia.org/wiki/Entropy_(information_theory)
+"""
+
+import sys
+import math
+r = range
```
Comment (cclauss) [inline maths/entropy.py:8]:
```suggestion
from string import ascii_lowercase
```

File: maths/entropy.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+def calculate_prob(my_dic_fir, my_dic_sec, path_dest):
+    """
+    This method takes path and two dict as argument
+    and than calculates entropy of them.
+    :param my_dic_fir:
+    :param my_dic_sec:
+    :param path_dest:
+    :return:
+    """
+    f = open(path_dest, mode="w+", encoding="utf-8")
```
Comment (cclauss) [inline maths/entropy.py:31]: Do not open the file until you need it.  Please move this down to line 47.

File: maths/entropy.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    my_alphas = create_alpha_array()  # create our alpha
+    # what is our total sum of probabilities.
+    values = my_dic_fir.values()
+    all_sum = sum(values)
```
Comment (cclauss) [inline maths/entropy.py:36]:
```suggestion
    all_sum = sum(my_dic_fir.values())
```

File: maths/entropy.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    # one length string
+    my_fir_sum = 0
+    # for each alpha we go in our dict and if it is in it we calculate entropy
+    for _ in my_alphas:
```
Comment (cclauss) [inline maths/entropy.py:41]: Only use _ as a variable name if you are never going to refer to it again.  In this case, please replace _ with `char`.

Or better yet, use `for key, value in first_dict.items():`.  The use of `my_` in variable names is not helpful.

File: maths/entropy.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    # write entropy to file
+    f.write("{0:.7f}".format(-1 * my_fir_sum))
+    f.write("\n")
+
+    # two len string
+    values = my_dic_sec.values()
+    all_sum = sum(values)  # time for two len probabilities sum
+    my_sec_sum = 0
+    # for each alpha (two in size) calculate entropy.
+    for _ in my_alphas:
```
Comment (cclauss) [inline maths/entropy.py:56]:
```suggestion
    for char0 in my_alphas:
```

File: maths/entropy.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        for __ in my_alphas:
```
Comment (cclauss) [inline maths/entropy.py:57]:
```suggestion
        for char1 in my_alphas:
```

File: maths/entropy.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    # write second entropy to the file
+    f.write("{0:.7f}".format(-1 * my_sec_sum))
+    f.write("\n")
+
+    # write the difference between them to the file
+    f.write("{0:.7f}".format((-1 * my_sec_sum) - (-1 * my_fir_sum)))
+    # f.write('\n')
+
+
+# 0x
```
Comment (cclauss) [inline maths/entropy.py:72]:
```suggestion
```

File: maths/entropy.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    And later on it open the file at this path and puts frequency on each dictionary
+    in first dictionary it puts one len strings.
+    in second dictionary it puts two len strings.
+    :param path_src:
+    :param path_dest:
+    :return:
+    """
+    my_dic_fir = {}
+    my_dic_sec = {}
+    f = open(path_src, mode="r", encoding="utf-8")
```
Comment (cclauss) [inline maths/entropy.py:85]:
```suggestion
    with open(path_src) as f:
```
This will automatically `f.close()`.

File: maths/entropy.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    text = f.read()
```
Comment (cclauss) [inline maths/entropy.py:86]:
```suggestion
        text = f.read()
```

File: maths/entropy.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    my_dic_fir[text[len(text) - 1]] = 1
+
+    # first case when we have space at start.
+    my_dic_sec[" " + text[0]] = 1
+    for _ in r(0, len(text) - 1):
+        sequence_fir = text[_]
+        sequence_sec = text[_ : _ + 2]
+        # continue
+        if sequence_fir not in my_dic_fir:
```
Comment (cclauss) [inline maths/entropy.py:95]: Please use https://docs.python.org/3/library/collections.html#collections.Counter

File: maths/entropy.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+            my_dic_fir[sequence_fir] = 1
+        else:
+            my_dic_fir[sequence_fir] += 1
+
+        if sequence_sec not in my_dic_sec:
```
Comment (cclauss) [inline maths/entropy.py:100]: Please use https://docs.python.org/3/library/collections.html#collections.Counter

File: maths/entropy.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+            my_dic_sec[sequence_sec] = 1
+        else:
+            my_dic_sec[sequence_sec] += 1
+
+    return my_dic_fir, my_dic_sec
+
+
+def main():
+    path_src = sys.argv[1]
+    path_dest = sys.argv[2]
```
Comment (cclauss) [inline maths/entropy.py:110]:
```suggestion
    try:
        _, path_src, path_dest = sys.argv
    except IndexError:
        print(f"Usage: {sys.argv[0]} path_src path_dest”)
```

File: maths/entropy.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    f.write("{0:.7f}".format(-1 * my_sec_sum))
+    f.write("\n")
+
+    # write the difference between them to the file
+    f.write("{0:.7f}".format((-1 * my_sec_sum) - (-1 * my_fir_sum)))
+    # f.write('\n')
+
+
+# 0x
+def open_file(path_src, path_dest):
```
Comment (cclauss) [inline maths/entropy.py:73]:
```suggestion
def read_dicts_from_file(path_src: str) -> dict, dict:
```
* [x] More self-documenting function name
* [x] Remove unused parameter
* [x] Python type hints as discussed in CONTRIBUTING.md

File: maths/entropy.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    """
+    This method take path and also two clear dict as argument
+    And later on it open the file at this path and puts frequency on each dictionary
+    in first dictionary it puts one len strings.
+    in second dictionary it puts two len strings.
+    :param path_src:
+    :param path_dest:
+    :return:
```
Comment (cclauss) [inline maths/entropy.py:81]:
```suggestion
    :return: dict, dict
```

**Comment by cclauss** [comment]
> This fail is not from my PR!

Yes but not surprising because this PR contains no tests.

**Comment by beqakd** [comment]
Will do changes!

**Comment by beqakd** [comment]
@cclauss  All changes done! Tests added. I think its cool now.

**Pull request merged**